### DISCUSSION TITLE
New features for OC4 event system

### DISCUSSION THREAD

**Discussion by v5mvidal** [opencart/opencart#10962] [Ideas]
@danielkerr, here are some points for analysis about the OpenCart 4 event system.

- Events should be automatically disabled when throwing PHP errors. This way we eliminate the risk of the store breaking, especially when upgrading.
- When an event throws errors and is disabled, it is important to store the error in a log so that it can be debugged.
- List all active and inactive events in a file, so it's easier to identify how many events are linked in a file. It would be important to display which extension added the event, this way we can separate native from non-native events.
- Sandbox mode for event. The basic idea is to generate event runtime and memory usage metrics. While these metrics are never exact due to a number of factors, it can help with performance issues.

Basically we have 3 implementations, one for error handling, another for event handling by file and extension, and the last one for performance.

I have other ideas, but I believe these are the most basic ones to start with.

I would like to receive opinions before making pull requests...

**Please! Don't talk here about how much you love ocmod or how much you want something similar.**

**Comment by felipoantonoff** [comment]
H, i liked the ideas and suggestions.
**Here are a few more suggestions for small changes to the Events screen:**
- Name of the Module that created the Event with the installation date
- On-screen log summarizing certain data for easy identification of file, line and other details
- Way to disable all third-party events for quick testing and validate if in case of a bug/conflict it is related to an event

I also think about analyzing if an Event can change PHP lines as a controller, model or library, if not, check ways to implement this feature to be able to have Modules that change native lines without touching the files directly as we did before in Opencart 3, because many modules depend on this feature.

A point on the performance issue is to add Cache, perhaps a simple Cache for certain Events, mainly those dealing with visual change and clearing these caches through the Panel.

**Reply by v5mvidal** [reply]
Name of module that created the event and date is important.

I don't see the need for such a detailed log, as the idea is to save all the details of the error generated by PHP when loading the broken event.

I don't know how the event cache would work, because the idea of the event is to be dynamic, that is, change in real time.

**Reply by v5mvidal** [reply]
Events work before or after loading the controller, model and view.

They can change any information in the input or output.

**Reply by felipoantonoff** [reply]
About the Logs, it would be more for adapting to themes or conflicts, but if the PHP error warning helps with that, it would already solve it.

Yes, Events has this detail of being more realtime, the idea I thought of the cache might just be a simple attribute in the creation of the Event that informs if its return will be cached or not, for example, to avoid the developer dealing with this little detail.

Maybe a simpler way to change the view would also be nice, maybe event variables that already deal with search and changes like XMLs operations I imagine, so the work of dealing with pulling the original view file, finding the lines to be modified and return the change, could be resolved by Events to make code more production-ready.

**Reply by felipoantonoff** [reply]
Yes, they can't touch a specific point in the PHP code, just before or after, I can't see a way for him to be able to do something in the middle of the process like change something in a PHP file other than its input or output.

Maybe a type of Event focused on changing the file and generating the cache, not being a trigger anymore, I just don't know if this is an idea for the future of this feature, but it would be good for View and for messing with PHP files and maintaining high performance for these use cases.

**Comment by TheCartpenter** [comment]
@v5mvidal: See this commit: https://github.com/opencart/opencart/issues/10985#issuecomment-1144796332 .
